contracts/contracts/coordination/BetaProgramInitiator.sol

```diff
@@ -0,0 +1,162 @@
+        IEncryptionAuthorizer accessController,
+        uint256 payment
+    );
+
+    event RequestCanceled(address indexed sender, uint256 indexed requestIndex);
+
+    event RequestExecuted(uint256 indexed requestIndex, uint256 indexed ritualId);
+
+    event FailedRequestRefunded(uint256 indexed requestIndex, uint256 refundAmount);
+
+    struct InitiationRequest {
+        address[] providers;
+        address authority;
+        uint32 duration;
+        IEncryptionAuthorizer accessController;
+        address sender;
+        uint32 ritualId;
+        uint256 payment;
+    }
+
+    uint32 public constant NO_RITUAL = type(uint32).max;
+
+    Coordinator public immutable coordinator;
+    IERC20 public immutable currency;
+    address public immutable executor; // TODO transferable role?
+
+    InitiationRequest[] public requests;
+
+    constructor(Coordinator _coordinator, address _executor) {
+        require(_executor != address(0), "Invalid parameters");
+        coordinator = _coordinator;
+        currency = coordinator.currency();
+        executor = _executor;
+    }
+
+    function getRequestsLength() external view returns (uint256) {
+        return requests.length;
+    }
+
+    function getProviders(uint256 requestIndex) external view returns (address[] memory) {
+        InitiationRequest storage request = requests[requestIndex];
+        return request.providers;
+    }
+
+    function registerInitiationRequest(
+        address[] calldata providers,
+        address authority,
+        uint32 duration,
+        IEncryptionAuthorizer accessController
+    ) external returns (uint256 requestIndex) {
+        uint256 ritualCost = coordinator.getRitualInitiationCost(providers, duration);
+
+        requestIndex = requests.length;
+        InitiationRequest storage request = requests.push();
+        request.providers = providers;
+        request.authority = authority;
+        request.duration = duration;
+        request.accessController = accessController;
+        request.sender = msg.sender;
+        request.ritualId = NO_RITUAL;
+        request.payment = ritualCost;
+
+        emit RequestRegistered(
+            msg.sender,
+            requestIndex,
+            providers,
+            authority,
+            duration,
+            accessController,
+            ritualCost
+        );
+        currency.safeTransferFrom(msg.sender, address(this), ritualCost);
+
+        return requestIndex;
+    }
+
+    function cancelInitiationRequest(uint256 requestIndex) external {
+        require(requestIndex < requests.length, "Non-existent request");
+        InitiationRequest storage request = requests[requestIndex];
+        address sender = request.sender;
+        require(msg.sender == sender || msg.sender == executor, "Not allowed to cancel");
+
+        uint256 ritualCost = request.payment;
+        require(request.ritualId == NO_RITUAL, "Request already executed");
+        require(ritualCost != 0, "Request canceled");
```

Comment:
```suggestion
        require(ritualCost != 0, "Request cancelled");
```

Reply:
To be honest, I find American "_canceled_" to more natural than "_cancelled_".

Reply:
Oh, ha! I didn't realize it was an American vs Canadian/British thing - in that case ignore hehe.